### ISSUE TITLE
ipahost: fails if sshpubkey is set to a list

### ISSUE THREAD

**Issue by okapia** [freeipa/ansible-freeipa#1074]
For ipahost, the sshpubkey (alias ipasshpubkey) key is documented to be of type list. If I set it to a string, it works and adds the single key to the host record in IPA. But if I set it to a list I get the error "host_mod: XXXX.XXX: invalid 'sshpubkey': invalid SSH public key". I have tried a variety of different keys and I don't think it is any particular key being bad. The same keys work running `ipa host-mod` and multiple `--sshpubkey` options.

I'm using this to add a few hosts that are not intended to be IPA clients so in some cases I don't even want the Kerberos principal. I'd just like my systems to get the host keys securely instead of relying on the usual TOFU procedure with ssh. It'd actually be nice if there was an option to use ssh-keyscan to populate the values initially. It'd probably be a mistake to update existing values in this manner because that would subvert the whole point. If they become out-of-date, the admin should need to manually invalidate them.

**Comment by rjeffman** [comment]
That is interesting, as this [behavior is tested](https://github.com/freeipa/ansible-freeipa/blob/b861a6185714bb56f3587ac482d37dd8e2ef2634/tests/host/test_host_sshpubkey.yml#L84-L94).

Can you provide more details on how you are invoking the playbook?

**Comment by okapia** [comment]
The error I get is `invalid 'sshpubkey': invalid SSH public key`

What I have is similar to the test case with an ipahost tag with `name` and `ipaadmin_password` set along with:

```
      sshpubkey:
      - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDblBchz4wxVOAKjFwUZAY7F2uV3G6UPlGSfWwVO7zB7Opw7X8SvftuQPQmy4cHSbWZC3BUBURFfrQzFhoR6XeKwR02QGzPcOdHM2t17tFQ+sqvncd7q4F7xEneUW4MDmn4sO8aAah8YlO7rLES/xDyRDjFajzdwFdg2aFAlm1IMq8a2v8WJXYeLBfz0R1T2dfNiNfH50s4r62xhVmT/P52gdDpejKzVxsZCEV9eIXXtg5G4xWBLOPjeEJiHIUSmeAcnv7SDnRGw2mGf+hK/wGYk+dt9/oIGOY2EI/FK1jcVuoHyoNqcpnXevJeq1FEVvffnIzcR+yWm1sYKtn7YXGB10s1+/mSpKjR2ZmHBDxh+crVNyquvEyMl4N4CpYv7rxgLnl/Kh53GEIf4FbqvH4IcBdthuUQaNwnZVs7ScMyOaw0/4ezs0YXRcZTtDjPHbzcSgeuviTc4oPyyrWF2jnrruIbowtkvHJo8YYHlu+H5tE1BIRPs0Og/PGRDvto58c=
      - ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBNqcHLnqgKxfJw7IRNnTzjHeSMS+kXqxsm0EYUDx87TE5YpEuO4KVz/2ZHTRbrPNEBNCtgVqs5b+av+DADFTIfE=
```

I've also tried substituting the literal values used in the test case with associated comments and those result in the same error. I'm more skeptical about whether the test case is using valid ssh keys - the second lacks the algorithm part, e.g. "ecdsa-sha2-nistp256" and both have e-mail address style comments at the end which would seem unusual for a host key.

I tried setting `ANSIBLE_STDOUT_CALLBACK=json` to capture more detail and somehow the array is ending up requoted into a string:

```
      "reverse": null,
      "sshpubkey": "['ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDblBchz4wxVOAKjFwUZAY7F2uV3G6UPlGSfWwVO7zB7Opw7X8SvftuQPQmy4cHSbWZC3BUBURFfrQzFhoR6XeKwR02QGzPcOdHM2t17tFQ+sqvncd7q4F7xEneUW4MDmn4sO8aAah8YlO7rLES/xDyRDjFajzdwFdg2aFAlm1IMq8a2v8WJXYeLBfz0R1T2dfNiNfH50s4r62xhVmT/P52gdDpejKzVxsZCEV9eIXXtg5G4xWBLOPjeEJiHIUSmeAcnv7SDnRGw2mGf+hK/wGYk+dt9/oIGOY2EI/FK1jcVuoHyoNqcpnXevJeq1FEVvffnIzcR+yWm1sYKtn7YXGB10s1+/mSpKjR2ZmHBDxh+crVNyquvEyMl4N4CpYv7rxgLnl/Kh53GEIf4FbqvH4IcBdthuUQaNwnZVs7ScMyOaw0/4ezs0YXRcZTtDjPHbzcSgeuviTc4oPyyrWF2jnrruIbowtkvHJo8YYHlu+H5tE1BIRPs0Og/PGRDvto58c=', 'ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBNqcHLnqgKxfJw7IRNnTzjHeSMS+kXqxsm0EYUDx87TE5YpEuO4KVz/2ZHTRbrPNEBNCtgVqs5b+av+DADFTIfE=']",
      "state": "present",
```
Using that alternate [ "...", "..." ] syntax doesn't help.

**Comment by okapia** [comment]
Digging further, I find that in `ipahost.py` in the ansible-freeipa package is the line
```
sshpubkey=dict(type="str", aliases=["ipasshpubkey"],
```

In the git sources, this has changed to:

```
sshpubkey=dict(type="list", elements="str", aliases=["ipasshpubkey"],
```

So perhaps this has been fixed in the git repo. I hope I don't need to wait for RHEL 10 to be able to use it.

**Comment by t-woerner** [comment]
This fix is part of ansible-freeipa 1.9.2, which is part of 8.8 and 9.2.

Alternatively there is an ansible-freeipa collection available in AnsibleAutomation Hub: redhat.rhel_idm (version 1.10.0)

Which RHEL release are you using?

**Comment by okapia** [comment]
It appears to be ansible-freeipa 1.8.3 on 8.7 and 9.1. If the fix comes with 9.2, that'll work for me.

**Comment by okapia** [comment]
I've upgraded to 9.2 and can confirm that it now works. Thanks for the help. Am closing the issue now.
